File: attestationpolicies/policies.go
// Copyright (c) 2021 Fraunhofer AISEC
// Fraunhofer-Gesellschaft zur Foerderung der angewandten Forschung e.V.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package attestationpolicies

import (
	"encoding/json"

	ar "github.com/Fraunhofer-AISEC/cmc/attestationreport"
	"github.com/robertkrimen/otto"
	"github.com/sirupsen/logrus"
)

var log = logrus.WithField("service", "policy-agent")

// JavaScriptValidator is a javascript implementation of the
// attestation report generic PolicyValidator interface
type JavaScriptValidator struct {
	policies []byte
}

// NewPolicyValidator creates a new JavaScriptValidator with custom policies.
// Custom policies are handed over as a byte array. This implementation
// accepts custom policies as javascript code. The javascript code
// can parse the VerificationResult in the variable 'json', i.e.:
//
//	var obj = JSON.parse(json);
//
// The javascript code must return a single boolean to indicate the
// success of the parsing. Logs can be output via: console.log()
// A very simple example of a custom Policy could look as follows:
//
//		var obj = JSON.parse(json);
//		var success = true;
//		if (obj.type != "Verification Result") {
//			console.log("Invalid type");
//			success = false;
//		}
//	    success
func NewPolicyValidator(policies []byte) *JavaScriptValidator {
	return &JavaScriptValidator{
		policies: policies,
	}
}

// Validate uses a javascript engine to validate the JavaScriptValidator's
// custom javascript policies against the verification result
func (p *JavaScriptValidator) Validate(result ar.VerificationResult) bool {

	log.Debugf("Validating custom javascript policies")

	vr, err := json.Marshal(result)
	if err != nil {
		log.Errorf("Failed to marshal verification result: %v", err)
		return false
	}

	// Create new javascript engine
	vm := otto.New()

	// Set variable json = vr
	vm.Set("json", string(vr))

	// Run javascript validation
	val, err := vm.Run(string(p.policies))
	if err != nil {
		log.Errorf("Failed run policy validation: %v", err)
		return false
	}

	// Retrieve result
	ok, err := val.ToBoolean()
	if err != nil {
		log.Errorf("Failed convert policy validation result: %v", err)
		return false
	}

	log.Debugf("Policy Validation: %v", ok)

	return ok
}

File: attestationpolicies/duktape/policies.c
// Copyright (c) 2021 Fraunhofer AISEC
// Fraunhofer-Gesellschaft zur Foerderung der angewandten Forschung e.V.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#define _POSIX_C_SOURCE 200809L
#include <stdio.h>
#include <sys/stat.h>
#include <stdbool.h>
#include <math.h>
#include <string.h>

#include "duktape.h"

// required because duktape does not provide I/O bindings
static duk_ret_t native_print(duk_context *ctx) {
    printf("%s\n", duk_to_string(ctx, 0));
    return 0;
}

char* Validate(uint8_t *ar, size_t ar_size, uint8_t *policies, size_t policies_size) {
    char ar_str[ar_size + 1];
    memcpy(ar_str, ar, ar_size);
    ar_str[ar_size] = '\0';

    char policies_str[policies_size + 1];
    memcpy(policies_str, policies, policies_size);
    policies_str[policies_size] = '\0';

    duk_context *ctx = duk_create_heap_default();

    // Optional: provide print() in JS
    duk_push_c_function(ctx, native_print, 1);
    duk_put_global_string(ctx, "print");

    // Push `json` variable for policy
    duk_push_global_object(ctx);
    duk_push_string(ctx, "json");
    duk_push_string(ctx, ar_str);
    duk_def_prop(ctx, -3, DUK_DEFPROP_HAVE_VALUE);

    // Evaluate policies
    if (duk_peval_string(ctx, policies_str) != 0) {
        const char *err = duk_safe_to_string(ctx, -1);
        fprintf(stderr, "Policy error: %s\n", err);
        duk_destroy_heap(ctx);
        return strdup("false");
    }

    // Convert result to string
    const char *result_str = duk_safe_to_string(ctx, -1);
    char *ret = strdup(result_str);

    duk_destroy_heap(ctx);
    return ret;
}
